pkg/provider/ip/provider.go

```diff
@@ -66,35 +75,59 @@ func (p *ipv4Provider) InitResource(instance ec2.EC2Instance) error {
 	nodeName := instance.Name()
 
 	eniManager := eni.NewENIManager(instance)
-	presentIPs, err := eniManager.InitResources(p.apiWrapper.EC2API)
+	presentIPs, presentPrefixes, err := eniManager.InitResources(p.apiWrapper.EC2API)
 	if err != nil {
 		return err
 	}
 
+	presentIPSet := map[string]struct{}{}
+	for _, ip := range presentIPs {
+		presentIPSet[ip] = struct{}{}
+	}
+
 	pods, err := p.apiWrapper.PodAPI.GetRunningPodsOnNode(nodeName)
 	if err != nil {
 		return err
 	}
 
-	podToResourceMap := map[string]string{}
+	podToResourceMap := map[string]pool.Resource{}
 	usedIPSet := map[string]struct{}{}
 	for _, pod := range pods {
 		annotation, present := pod.Annotations[config.ResourceNameIPAddress]
 		if !present {
 			continue
 		}
-		podToResourceMap[string(pod.UID)] = annotation
-		usedIPSet[annotation] = struct{}{}
+		// Only mark pod as used if it's secondary IP
+		if _, found := presentIPSet[annotation]; found {
+			podToResourceMap[string(pod.UID)] = pool.Resource{GroupID: annotation, ResourceID: annotation}
+			usedIPSet[annotation] = struct{}{}
+		}
 	}
 
-	warmResources := difference(presentIPs, usedIPSet)
+	warmIPs := difference(presentIPs, usedIPSet)
+	warmResources := make(map[string][]pool.Resource, len(presentIPs))
+	for _, ip := range warmIPs {
+		warmResources[ip] = append(warmResources[ip], pool.Resource{GroupID: ip, ResourceID: ip})
+	}
 
-	nodeCapacity := getCapacity(instance.Type(), instance.Os())
-	resourcePool := pool.NewResourcePool(p.log.WithName("ipv4 resource pool").
-		WithValues("node name", instance.Name()), p.config, podToResourceMap,
-		warmResources, instance.Name(), nodeCapacity)
+	nodeCapacity := getCapacity(instance.Type(), instance.Os()) - len(presentPrefixes)
 
-	p.putInstanceProviderAndPool(nodeName, resourcePool, eniManager)
+	// Set warm pool config to 0 if PD is enabled
+	ipV4WarmPoolConfig := p.config
+	isPDEnabled := p.conditions.IsWindowsPrefixDelegationEnabled()
+	if isPDEnabled {
+		ipV4WarmPoolConfig = &config.WarmPoolConfig{
+			DesiredSize:  0,
```

Comment:
DesiredSize is 0? Is this intended?

Reply:
Ahh yes, this is intended because for secondary IP provider, once it sees PD enabled, it sets its own secondary IP pool to `desiredSize` of 0, effectively draining the pool.